api/src/main/java/net/md_5/bungee/api/ServerConnectRequest.java

```diff
@@ -0,0 +1,107 @@
+package net.md_5.bungee.api;
+
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NonNull;
+import net.md_5.bungee.api.config.ServerInfo;
+import net.md_5.bungee.api.event.ServerConnectEvent;
+
+/**
+ * A request to connect a server
+ */
+@Getter
+@Builder
+public class ServerConnectRequest
+{
+
+    public interface Callback
+    {
+
+        void done(Result result, Throwable error);
+
+        enum Result {
+
+            /**
+             * ServerConnectEvent to the new server was canceled.
+             */
+            EVENT_CANCEL,
+            /**
+             * Already connected to target server.
+             */
+            ALREADY_CONNECTED,
+            /**
+             * Already connecting to target server.
+             */
+            ALREADY_CONNECTING,
+            /**
+             * Successfully connected to server.
+             */
+            SUCCESS,
+            /**
+             * Connection failed, error can be accessed from callback method handle.
+             */
+            FAIL
+        }
+    }
+
+    public static class LegacyCallback implements net.md_5.bungee.api.Callback<Boolean>, Callback
+    {
+
+        private final net.md_5.bungee.api.Callback<Boolean> callback;
+
+        public LegacyCallback(net.md_5.bungee.api.Callback<Boolean> callback)
+        {
+            this.callback = callback;
+        }
+
+        @Override
+        public final void done(Result result, Throwable error)
+        {
+            done( ( result == Result.SUCCESS ) ? Boolean.TRUE : Boolean.FALSE, error );
+        }
+
+        @Override
+        public void done(Boolean result, Throwable error)
+        {
+            callback.done( result, error );
+        }
+    }
+
+    /**
+     * Target server to connect to.
+     */
+    @NonNull
+    private final ServerInfo target;
+    /**
+     * Reason for connecting to server.
+     */
+    @NonNull
+    private final ServerConnectEvent.Reason reason;
+    /**
+     * Callback sent back from request.
+     */
+    private final ServerConnectRequest.Callback callback;
```

Comment:
`net.md_5.bungee.api.Callback<Result>` ?

Reply:
No because that would break plugins using the old Callback<Boolean>

Reply:
I chose not to do this because there will be a discrepancy between the old UserConnection#connect method with the Callback<Boolean> arg and it may confuse plugin developers.

Reply:
On second thought, you're right and I have gone ahead and changed it to that removing some unnecessary bulk.